### DISCUSSION TITLE
How to build application with libcurl static linking on ubuntu

### DISCUSSION THREAD

**Discussion by Bab95** [curl/curl#10302] [Q&A]
How to define libcurl dependency in cmake file in ubuntu? 

/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-url.o): In function `Curl_disconnect':
(.text+0xb4c): undefined reference to `idn2_free'
(.text+0xb68): undefined reference to `idn2_free'
(.text+0xb84): undefined reference to `idn2_free'
(.text+0xba0): undefined reference to `idn2_free'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-url.o): In function `Curl_connect':
(.text+0x4e6b): undefined reference to `idn2_free'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-url.o):(.text+0x4e87): more undefined references to `idn2_free' follow
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-url.o): In function `fix_hostname.isra.6':
(.text.unlikely+0x2e4): undefined reference to `idn2_check_version'
(.text.unlikely+0x301): undefined reference to `idn2_lookup_ul'
(.text.unlikely+0x34b): undefined reference to `idn2_strerror'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-version.o): In function `curl_version':
(.text+0xaa): undefined reference to `zlibVersion'
(.text+0xd6): undefined reference to `idn2_check_version'
(.text+0xe2): undefined reference to `idn2_check_version'
(.text+0x106): undefined reference to `psl_get_version'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-version.o): In function `curl_version_info':
(.text+0x1d6): undefined reference to `zlibVersion'
(.text+0x1e9): undefined reference to `idn2_check_version'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `before_frame_send':
(.text+0x38): undefined reference to `nghttp2_session_get_stream_user_data'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `on_invalid_frame_recv':
(.text+0x58): undefined reference to `nghttp2_session_get_stream_user_data'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `on_header':
(.text+0xac): undefined reference to `nghttp2_session_get_stream_user_data'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `on_begin_headers':
(.text+0x3e8): undefined reference to `nghttp2_session_get_stream_user_data'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `data_source_read_callback':
(.text+0x44a): undefined reference to `nghttp2_session_get_stream_user_data'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o):(.text+0x52f): more undefined references to `nghttp2_session_get_stream_user_data' follow
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `http2_perform_getsock':
(.text+0x634): undefined reference to `nghttp2_session_want_write'
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/libcurl.a(libcurl_la-http2.o): In function `h2_pri_spec':
(.text+0x695): undefined reference to `nghttp2_priority_spec_init'

I am getting this error while trying to libcurl.a with the application? Does anybody know what could be the problem?

**Comment by jzakrzewski** [comment]
If you just link 'libcurl.a', it's wrong. When linking static libraries, you must also provide all the dependencies - this how static libraries work. You have 3 options:
* figure out all dependencies and link them correctly
* I'd assume pkconfig would have the correct dependencies
* think again, if you really need static linking

**Comment by Bab95** [comment]
@jzakrzewski  where can i get the list of other dependenices required by libcurl.

**Comment by jzakrzewski** [comment]
I am not an expert here. I know there should be something like libcurl.pc installed somewhere - it's a PKGCONFIG file that describes the include directories, and stuff related to linking.

**Comment by andy5995** [comment]
@Bab95 Do you still need help with this? Is your project on GitHub, or visible somewhere that you can give me a link?

Have a look at this custom cmake file that's used to add libcurl to a project and maybe you'll get some clues:

https://github.com/tommccallum/oauth2/blob/ac1fee7f732bb5b594ee048e7d147715183e0fb0/cpp/cmake/FindCurlCustom.cmake#L5

**Comment by Bab95** [comment]
@andy5995 I was able to link static curl on ubuntu. Closing this thread now.

**Reply by UnshiftedEarth** [reply]
I need help statically linking curl on ubuntu. I have tried everything and still cannot get the dependencies resolved. Can you help me?

**Reply by Bab95** [reply]
cmake_minimum_required(VERSION 3.5)

project(ProjName)

set(CMAKE_PROJECT_TARGET ProjName)

add_definitions (-DPLATFORM_UNIX)

find_path(CURL_INCLUDE_DIR NAMES curl PATHS /usr/local/curl/include
                                            NO_DEFAULT_PATH)
include_directories(${CURL_INCLUDE_DIR})


add_executable(${CMAKE_PROJECT_TARGET} main.cpp)

find_library(CURL_LIB NAMES libcurl.a PATHS /usr/local/curl/lib
                                           NO_DEFAULT_PATH)

target_link_libraries(${CMAKE_PROJECT_TARGET} stdc++fs)
target_link_libraries(${CMAKE_PROJECT_TARGET} ${CURL_LIB} z)
target_link_libraries(${CMAKE_PROJECT_TARGET} libssl.a)



your cmake build file should look like this. assuming I have libssl.a and libcurl.a was build with openssl support.

**Comment by dfandrich** [comment]
Adding to somebody else's year-old closed thread isn't going to help. Open a
new one and tell us exactly what you've tried, what errors messages you see,
what your build environment is, and which of the FAQs and documents you've read
that did not help with solving the problem.
